langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModelIT.java

```diff
@@ -413,4 +424,153 @@ public void onError(Throwable error) {
+        assertThat(request.temperature()).isEqualTo(temperature);
+        assertThat(request.topP()).isEqualTo(topP);
+        assertThat(request.maxTokens()).isEqualTo(maxTokens);
+        assertThat(request.messages()).containsExactly(userMessage);
+        assertThat(request.toolSpecifications()).containsExactly(toolSpecification);
+
+        ChatModelResponse response = responseReference.get();
+        assertThat(response.id()).isNotBlank();
+        assertThat(response.model()).isNotBlank();
+        assertThat(response.tokenUsage().inputTokenCount()).isGreaterThan(0);
+        assertThat(response.tokenUsage().outputTokenCount()).isGreaterThan(0);
+        assertThat(response.tokenUsage().totalTokenCount()).isGreaterThan(0);
+        assertThat(response.finishReason()).isNotNull();
+        assertThat(response.aiMessage()).isEqualTo(aiMessage);
+    }
+
+    @Test
+    void should_listen_error() throws Exception {
+
+        // given
+        String wrongApiKey = "banana";
+
+        AtomicReference<ChatModelRequest> requestReference = new AtomicReference<>();
+        AtomicReference<Throwable> errorReference = new AtomicReference<>();
+
+        ChatModelListener listener = new ChatModelListener() {
+
+            @Override
+            public void onRequest(ChatModelRequestContext requestContext) {
+                requestReference.set(requestContext.request());
+                requestContext.attributes().put("id", "12345");
+            }
+
+            @Override
+            public void onResponse(ChatModelResponseContext responseContext) {
+                fail("onResponse() must not be called");
+            }
+
+            @Override
+            public void onError(ChatModelErrorContext errorContext) {
+                errorReference.set(errorContext.error());
+                assertThat(errorContext.request()).isSameAs(requestReference.get());
+                assertThat(errorContext.partialResponse()).isNull(); // can be non-null if it fails in the middle of streaming
+                assertThat(errorContext.attributes().get("id")).isEqualTo("12345");
+            }
+        };
+
+        StreamingChatLanguageModel model = AzureOpenAiStreamingChatModel.builder()
+            .endpoint(System.getenv("AZURE_OPENAI_ENDPOINT"))
+            .apiKey(wrongApiKey)
+            .logRequestsAndResponses(true)
+            .listeners(singletonList(listener))
+            .build();
+
+        String userMessage = "this message will fail";
+
+        CompletableFuture<String> future = new CompletableFuture<>();
+        StreamingResponseHandler<AiMessage> handler = new StreamingResponseHandler<AiMessage>() {
+
+            @Override
+            public void onNext(String token) {
+                fail("onNext() must not be called");
+            }
+
+            @Override
+            public void onError(Throwable error) {
+                fail("OnError() must not be called");
```

Comment:
Hi @qiaoleiatms, I've jsut noticed that this test (and implementation in `AzureOpenAiStreamingChatModel`) is different from what has been implemented in other models (e.g. `OpenAiStreamingChatModel`).

In case of errors, `onError()` should be called, not `onComplete()`. Could you please check?

BTW, I have moved tests to `AzureOpenAiStreamingChatModelListenerIT` and disabled `should_listen_error` untill this discrepancy is fixed.

Thanks!

Reply:
@langchain4j that is because while exception happen, the HttpResponseException will be treated as a complete response from the original implementation 

https://github.com/langchain4j/langchain4j/blob/a2004b689a5a010cfef20fadef4fff6d14efaca2/langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModel.java#L329

@jdubois could you please help to confirm the discrepency between the implementation of OpenAiStreamingChatModel? it treats all the exception as error

Reply:
I'm not totally sure.
For the "content_filter", the explanation is that on Azure OpenAI we have some specific filters for removing violence, sexual content, etc. Those are not part of the LLM, but are added on top of it. So they would not be an "error" (=the service could not process the request), but as a response that "completed" (=it worked, but it was filtered out as the user requested something that shouldn't be generated). I don't think the other services have this, so it's indeed specific to Azure OpenAI.

Reply:
I am not sure about harmful content, but in this specific case the problem is wrong API key, which IMHO should definitely result in an exception and `onError()` being called instead of `onComplete()`

Reply:
@langchain4j yeah, get your point, but I believe this is handled by the azure openai sdk(=complete the request even it's an API key error) 
@jdubois how do you think about it?

Reply:
@qiaoleiatms

> I believe this is handled by the azure openai sdk(=complete the request even it's an API key error)

Not really, SDK is throwing this exception:
`com.azure.core.exception.ClientAuthenticationException: Status code 401, "{"error":{"code":"401","message":"Access denied due to invalid subscription key or wrong API endpoint. Make sure to provide a valid key for an active subscription and use a correct regional API endpoint for your resource."}}"`